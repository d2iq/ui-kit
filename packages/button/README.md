# Buttons

Buttons are used for action items. Some examples include Create, Save, Delete. Buttons make it obvious to the user that they can take an action.

Do not use buttons as navigational elements. Instead, use anchors because it takes the user to a new page and is not associated with an action.

Each page will likely have one primary button. Any remaining calls-to-action are represented as secondary buttons.

## Button states and types

### Types

**Primary**: Primary call to action button. Should stand out above all other actions. Do not put two primary buttons next to each other. Examples include Save, Deploy.

**Secondary**: Most buttons on a page are likely this style.

**Danger**: Destructive buttons are used to highlight something dangerous that can't be undone e.g. "Delete".

**Icon only**: Use when an icon is enough to represent the action e.g. "Create a Service". Should be accompanied by a tooltip.

**Icon with label**: If there is space and we feel an icon by itself is not obvious enough e.g. "Kill Task"

### States

**Disabled**: Disable a button when the action is not applicable but we want to show the user that it may be possible. Disabled buttons should have a tooltip.

**Processing**: When user clicks or taps a button and an API request is initiated we should provide feedback so they know their action was received.

**Hidden:** Hide a button when there is no way a user can interact with the button.

## Using buttons as links

Sometimes there is a button that is actually a link to somewhere else. If the appropriate props are set, the button is rendered as an anchor tag, but it is styled like a button.

It is not recommended to use a button as a link when it triggers a dropdown. It would be frustrating if a user clicked a button and expected to see a dropdown appear, but they are taken to a new page.

## Button content

Button labels tell users what will happen when they click the button. Use verbs that describe the action. Acceptable labels include:

- New \$object
- Create \$object
- Edit \$object
- Save \$object
- Delete \$object
- Deploy, Scale, Restart

Use sentence-style capitalization (only the first word in a phrase and any proper nouns capitalized) and no more than three words for button labels.

For Sets of Buttons, use specific labels, such as Save or Cancel, instead of using Yes and No. This is particularly helpful when the user is confirming an action.

## ResetButton

A `ResetButton` component is essentially an unstyled HTML `<button>` element. It can be used for an element that is semantically a button, but is not intended to look like a button.

## Feedback

Buttons should have hover, active, focus, disabled states.

When a user interacts with a button, either via click or tap, and an action is being completed, e.g. an API call, the button should change to disabled and provide feedback via the label to the user that it received the interaction and is busy performing the action e.g. “Saving…”

## Tooltips

Buttons may or may not require tooltips to help communicate their intent.

Icon buttons that do not have labels should always have a tooltip so the user knows what will happen when they click it.

Disabled buttons should always have a tooltip to explain why it is disabled so the user is aware.
