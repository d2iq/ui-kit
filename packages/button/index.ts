export { default as PrimaryButton } from "./components/PrimaryButton";
export { default as SecondaryButton } from "./components/SecondaryButton";
export { default as StandardButton } from "./components/StandardButton";
export { default as SuccessButton } from "./components/SuccessButton";
export { default as DangerButton } from "./components/DangerButton";

export { default as PrimaryDropdownButton } from "./components/PrimaryDropdownButton";
export { default as SecondaryDropdownButton } from "./components/SecondaryDropdownButton";
export { default as StandardDropdownButton } from "./components/StandardDropdownButton";
export { default as SuccessDropdownButton } from "./components/SuccessDropdownButton";
export { default as DangerDropdownButton } from "./components/DangerDropdownButton";

export { default as ResetButton } from "./components/ResetButton";
