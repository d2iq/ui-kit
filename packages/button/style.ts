import { css } from "@emotion/css";
import {
  borderRadiusSmall,
  borderRadiusDefault,
  themeBrandPrimary,
  themeSuccess,
  themeError,
  themeTextColorInteractive,
  themeTextColorInteractiveInverted,
  themeTextColorPrimary,
  themeTextColorPrimaryInverted,
  themeBgDisabled,
  themeBgDisabledInverted,
  themeTextColorDisabled,
  themeBgNeutral,
  themeBgNeutralInverted,
  themeBrandPrimaryInverted,
  themeSuccessInverted,
  themeErrorInverted,
  themeTextColorDisabledInverted
} from "../design-tokens/build/js/designTokens";
import { darken, pickReadableTextColor } from "../shared/styles/color";
import { ButtonAppearances } from "./components/ButtonBase";
import { tintContent } from "../shared/styles/styleUtils";
import getCSSVarValue from "../utilities/getCSSVarValue";

export const buttonPadding = {
  vert: "10px",
  horiz: "18px"
};

const getHoverColor = color => darken(color, 1);
const getActiveColor = color => darken(color, 2);

export const filledButton = (
  baseColor: string,
  hoverColor: string,
  activeColor: string,
  isInverse?: boolean
) => {
  const contentColor = isInverse
    ? themeTextColorPrimaryInverted
    : pickReadableTextColor(
        baseColor,
        getCSSVarValue(themeTextColorPrimary),
        getCSSVarValue(themeTextColorPrimaryInverted)
      );

  return css`
    ${tintContent(contentColor)};
    background-color: ${baseColor};
    border-radius: ${borderRadiusSmall};
    padding: ${buttonPadding.vert} ${buttonPadding.horiz};

    &[href],
    &[href]:visited {
      ${tintContent(contentColor)};
    }

    &:hover {
      background-color: ${hoverColor};
    }

    &:active {
      background-color: ${activeColor};
    }
  `;
};

const mutedButton = css`
  ${tintContent(themeTextColorDisabled)};
  cursor: default;
  pointer-events: none;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabled)};
  }
`;

const inverseMutedButton = css`
  ${tintContent(themeTextColorDisabledInverted)};
  cursor: default;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabledInverted)};
  }
`;

const mutedFilledButton = css`
  background-color: ${themeBgDisabled};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabled};
  }
`;

const inverseMutedFilledButton = css`
  background-color: ${themeBgDisabledInverted};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabledInverted};
  }
`;

export const processingTextStyle = css`
  &:after {
    content: "...";
  }
`;

export const buttonBase = css`
  cursor: pointer;
  display: inline-block;
  outline: none;
  text-decoration: none;
`;

export const buttonContent = css`
  outline: none;
`;

export const fullWidthButton = css`
  text-align: center;
  width: 100%;
`;

export const focusStyles = (
  bgColor: string,
  borderColor: string = bgColor
) => css`
  outline: none;
  position: relative;

  &:focus {
    background-color: ${bgColor};

    &:after {
      border: 2px solid ${borderColor};
      border-radius: ${borderRadiusDefault};
      bottom: -3px;
      content: "";
      left: -3px;
      position: absolute;
      right: -3px;
      top: -3px;
    }
  }
`;

export const focusStyleByAppearance = (appearance, isInverse) => {
  switch (appearance) {
    case "primary":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimary)));
    case "secondary":
      return css`
        ${isInverse
          ? focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
            )
          : focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractive))
            )};
        &:focus {
          &:after {
            border-radius: 0;
            border-left-width: 0;
            border-right-width: 0;
            border-top-width: 0;
          }
        }
      `;
    case "standard":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBgNeutralInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBgNeutral)));
    case "success":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeSuccessInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeSuccess)));
    case "danger":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeErrorInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeError)));
    default:
      return "";
  }
};

export const button = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimary),
        getHoverColor(getCSSVarValue(themeBrandPrimary)),
        getActiveColor(getCSSVarValue(themeBrandPrimary))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractive)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractive)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutral),
        getHoverColor(getCSSVarValue(themeBgNeutral)),
        getActiveColor(getCSSVarValue(themeBgNeutral))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccess),
        getHoverColor(getCSSVarValue(themeSuccess)),
        getActiveColor(getCSSVarValue(themeSuccess))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeError),
        getHoverColor(getCSSVarValue(themeError)),
        getActiveColor(getCSSVarValue(themeError))
      );
    default:
      return "";
  }
};

export const buttonInverse = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimaryInverted),
        getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)),
        getActiveColor(getCSSVarValue(themeBrandPrimaryInverted))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractiveInverted)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractiveInverted)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutralInverted),
        getHoverColor(getCSSVarValue(themeBgNeutralInverted)),
        getActiveColor(getCSSVarValue(themeBgNeutralInverted))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccessInverted),
        getHoverColor(getCSSVarValue(themeSuccessInverted)),
        getActiveColor(getCSSVarValue(themeSuccessInverted))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeErrorInverted),
        getHoverColor(getCSSVarValue(themeErrorInverted)),
        getActiveColor(getCSSVarValue(themeErrorInverted))
      );
    default:
      return "";
  }
};

export const getMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${mutedButton};
        ${mutedFilledButton};
      `
    : mutedButton;
};

export const getInverseMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${inverseMutedButton};
        ${inverseMutedFilledButton};
      `
    : inverseMutedButton;
};

export const pointerCursor = css`
  cursor: pointer;
`;

// Replicates default browser focus ring styles.
//
// The media query targets Webkit browsers, which can
// more accurately replicate the native focus ring style
export const keyboardFocus = css`
  > div:focus {
    outline: none;
  }

  &:focus > div {
    outline-color: Highlight;
    outline-width: thin;

    @media (-webkit-min-device-pixel-ratio: 0) {
      outline-color: -webkit-focus-ring-color;
      outline-style: auto;
      outline-width: unset;
    }
  }
`;
