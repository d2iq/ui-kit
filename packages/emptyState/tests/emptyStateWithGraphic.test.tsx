import React from "react";

import { EmptyStateWithGraphic } from "..";
import { render } from "enzyme";
import toJSON from "enzyme-to-json";
import { PrimaryButton, SecondaryButton } from "../../button";

describe("EmptyStateWithGraphic", () => {
  it("default", () => {
    expect(
      toJSON(
        render(
          <EmptyStateWithGraphic
            graphicSrc="http://placehold.it/350x150"
            heading="Heading"
          >
            Body content
          </EmptyStateWithGraphic>
        )
      )
    ).toMatchSnapshot();
  });
  it("renders with primary and secondary actions", () => {
    expect(
      toJSON(
        render(
          <EmptyStateWithGraphic
            graphicSrc="http://placehold.it/350x150"
            heading="Heading"
            primaryAction={<PrimaryButton>Primary</PrimaryButton>}
            secondaryAction={<SecondaryButton>Secondary</SecondaryButton>}
          >
            Body content
          </EmptyStateWithGraphic>
        )
      )
    ).toMatchSnapshot();
  });
});
