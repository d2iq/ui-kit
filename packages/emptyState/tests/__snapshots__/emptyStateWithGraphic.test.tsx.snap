// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyStateWithGraphic default 1`] = `
<div
  data-cy="emptyStateWithGraphic"
>
  <div
    class="css-1gikt66"
    data-cy="spacingBox"
  >
    <img
      alt=""
      class="css-1jv97ae"
      role="presentation"
      src="http://placehold.it/350x150"
    />
  </div>
  <div
    class="css-so2jx9"
    data-cy="spacingBox"
  >
    <div
      class="css-e463k1"
      data-cy="spacingBox"
    >
      <h2
        class="css-zf0nd2"
        data-cy="headingText2"
      >
        Heading
      </h2>
    </div>
    <div
      class="css-1j4gpfm"
    >
      Body content
    </div>
  </div>
</div>
`;

exports[`EmptyStateWithGraphic renders with primary and secondary actions 1`] = `
<div
  data-cy="emptyStateWithGraphic"
>
  <div
    class="css-1gikt66"
    data-cy="spacingBox"
  >
    <img
      alt=""
      class="css-1jv97ae"
      role="presentation"
      src="http://placehold.it/350x150"
    />
  </div>
  <div
    class="css-1gikt66"
    data-cy="spacingBox"
  >
    <div
      class="css-e463k1"
      data-cy="spacingBox"
    >
      <h2
        class="css-zf0nd2"
        data-cy="headingText2"
      >
        Heading
      </h2>
    </div>
    <div
      class="css-1j4gpfm"
    >
      Body content
    </div>
  </div>
  <div
    class="css-8h0s7x"
    data-cy="flex"
  >
    <div
      class="css-149ndna"
      data-cy="flexItem"
    >
      <button
        class="css-67j5fi"
        data-cy="secondaryButton"
        tabindex="0"
        type="button"
      >
        <span
          class=""
        >
          Secondary
        </span>
      </button>
    </div>
    <div
      class="css-149ndna"
      data-cy="flexItem"
    >
      <button
        class="css-s9tdxz"
        data-cy="primaryButton"
        tabindex="0"
        type="button"
      >
        <span
          class=""
        >
          Primary
        </span>
      </button>
    </div>
  </div>
</div>
`;
